### ISSUE TITLE
Timing of adding a new folder and then putting files in it

### ISSUE THREAD

**Issue by phil-davis** [owncloud/client#3543]
Scenario A:
I add a new folder to my user on the server, intending to put a heap of files in it. After adding the folder I go into the folder and press the Upload button and start selecting some files and they start uploading. The end result might be a folder with 150MB of content.

My client has a confirmation limit of 100MB. But it takes time for the seeding of the 150MB of files on the server to happen (including time for me as the user to click the various buttons to select the files and even start the upload). So there is a window of time when the folder is empty and then where it has 1, 2, 3, 4... files in it less than the 100MB confirmation limit. If the client wakes up and talks to the server in that window, it sees a new folder with < confirmation limit size files in it. So it auto-downloads it.

A few minutes later the folder size is > confirmation limit. But the folder has already been auto-downloaded. So all the recent new files in the folder also auto-download.

Scenario B:
This might even apply to a situation with 2 clients - userA client1 creates the folder locally with a heap of files in it. They start uploading slowly. userA client2 sees the new folder while it is still small and auto-synches it. The folder gets much bigger as client1 continues to upload (slowly) and actually should not have been auto-synced to client2.

Scenario A really happens, Scenario B I have not tried but I guess it could happen.

I am not sure how to handle this.
In scenario B, client1 could tell the server the total size of everything that is coming in the new folder, and the server could use that somehow as the "pending folder size" to work out the confirmation limit with the client. (or not make the new folder visible to any other clients until client1 had completely uploaded it)
In Scenario A it is difficult. The time delay is due to user interaction with the web interface on the server. How long to wait to consider the new folder has been "seeded" with its initial file set? What happens when the user gets interrupted - phone call, coffee, internet down - and so the timer expires and the server thinks the initial folder "seeding" is done. And a heap of other cases...

For this new folder stuff, there needs to be some way that a new folder can be first created, then filled with files, then declared "ready" on the server.

Anyway, this is just a brain dump of the scenarios for discussion.

**Comment by MTRichards** [comment]
Hm. This is an interesting point about how this can sneak large folders by the limit. The real intent of the folder limit was to prevent extremely large folders like Windows Network Drives and FTP servers, from syncing all of the content everywhere by default - we get that complaint a lot. But this particular problem is interesting.

In general, there should be a delay after a new folder is created on the server which *should* allow you to copy in some files. This delay is based on the client poll interval and is just how the client works now. This could be anywhere from 0 to 30 seconds, depending on when the folder on the server updates arrive, but on average it is a measurable, if short, delay.

Given this delay already happens, I am not certain a longer delay could work. One option would be to go the other way, and have the server set the sync default as well. That way a folder when shared could tell the client whether or not sync of that folder is allowed. This, of course, would require an entirely new internal architecture for handling this meta data, and client updates to understand what to do about it.

To keep it simple, what if we just allowed you to set the comparison on the small side of a sync to an = on the small side? Then you could set a desktop client to not sync an empty folder at all? That means that nothing would sync by default, but does that work?

**Comment by moscicki** [comment]
Before sync client starts downloading any content it PROPFINDs the folder. This is the moment to get the size/quota information and use it. If more stuff is added meanwhile this will be only visible in the next PROPFIND cycle with updated size/quota information. So everytime sync client is discovering new content it may also get most recent size information and act accordingly. Would it make sense for your scenario?

The scenario that escapes this logic is the following: client starts downloading 1000 small files one-by-one. In the meantime a new version of the last file is uploaded by someone and this new version is a large file. This size difference would not be known by the client in advance.

**Comment by jospoortvliet** [comment]
BTW this happened at the conference in Berlin - I added a folder with about 4 GB of 5MB images and 3 GB of ~100 mb videos to our shared S3. Everybody who was online with their client got all 7GB of data, despite the client setting of 500 MB default.

And that is when I moved that folder into an ownCloud-synced location in one move. I could have added the images directly from the camera's slow SD card to the ownCloud-synced folder, working around a possible solution involving the client telling the server in advance how big a folder is going to be once fully uploaded (and other clients using that info to decide to auto-sync it or not).

A solution would be complicated and require some vague algorithms. Something like "stop syncing a folder which was added in the last hour and has grown beyond the limit and ask for confirmation first".

**Comment by phil-davis** [comment]
Or some way for a client to tell the server that a new folder and contents upload is to be considered a "transaction" and that the server should not advertise the existence of the folder or any of its files until the whole upload is complete. That way the whole folder and files can "suddenly" become available on the server when the upload finishes, rather than the content dribbling into existence.

**Comment by moscicki** [comment]
You are certainly right but the problem is that there is no such a thing as a transaction for directories (and neither for files). Sync client has to guess such a transaction boundary. In principle the sync client could postpone the upload of files if there is activity in the last n-seconds in this directory (same as it does for files which are being updated, e.g. https://github.com/cernbox/smashbox/blob/master/lib/test_slowwrite.py). However, there would always be a way of going around that: imagine, that I add a folder which is initially empty and which is synced to your client. After few minutes I add 100GB worth of files to it. The directory is already added on your side and will be synced as usual.

And after all someone may argue that the faster sync client starts the upload the better. 

I am not sure if there is any easy/sensible solution to this problem.

kuba

--

**Comment by phil-davis** [comment]
Yes, the server API would have to support some sort of "multi-file/folder upload transaction boundaries" thing, and keep the progressively uploaded files for a "transaction" hidden away somewhere until the uploading client sends a "commit".
And then what to do if the uploading client goes off-line part way through? And comes back online? Or some other client starts uploading a matching folder name, and the 2 will conflict when each of their "transactions" are completed? or...
It could get complex to know what to do with a "left behind part-uploaded folder".
And I wonder if it anything like this will be worth all the effort anyhow.

**Comment by jospoortvliet** [comment]
@phil-davis to quote myself from https://github.com/owncloud/client/issues/3543#issuecomment-138927013 :
I could have added the images directly from the camera's slow SD card to the ownCloud-synced folder, working around a possible solution involving the client telling the server in advance how big a folder is going to be once fully uploaded (and other clients using that info to decide to auto-sync it or not).

Transactions wouldn't always work either. And as Kuba said, you slow down the uploading or downloading or both this way. 

But this is going to result in a slow but steady stream of bug reports, I think. Unless @ogoffart comes up with a very clever solution it might be best to turn the feature in a "always ask to add new folders" checkbox. That will work consistently and without surprises.

I do think you can come with a ~95% solution, using some combination of things including telling the server how big a folder is before starting to upload and giving folders a bit of time to grow before telling the server their size but as @phil-davis said - not sure if it is worth the effort. That's a call for either "whoever wants to put his/her time in it" or @MTRichards ;-)

**Comment by dragotin** [comment]
I think this can only be solved on the server. There could be a functionality called "bulk upload directory" or so that stays hidden until the user explicitly makes it visible in the normal file collection. That way the user could create it, upload all the files, and finally make it visible. 

This could probably be implemented in an app.

The only thing that the client could do is stopping to sync once an existing dir exceeds the size limit and ask again. I am not sure if that would be a good solution because that would be confusing for most users.

**Comment by jospoortvliet** [comment]
@dragotin I'm not sure how the server can help, to be honest.

Say I create a new subfolder in my pictures folder which is synced with the client and shared with you and I start copying photos from my SD card into the folder... Each gets synced, one by one. Your client grabs the folder. Then the first image. Then the next.

With a slow SD card and 1GB of images, we got a scenario almost impossible to reliably detect. You can only do heuristics on this one and accept you do it wrong occasionally.

**Comment by hodyroff** [comment]
Reading the original request it seems that the correct way to politely add a large folder of such files is to first create it in your private area and then share it when you have it filled ... so the user who creates has it fully in his hands ... Don't see a bug here. Will continue to take it into consideration for handling of new folders, selective sync, etc.

**Comment by phil-davis** [comment]
Yes correc t, there is no bug. The system works as designed. This is an enhancement discussion. In the 18 months since creating this issue, I haven't thought of any great new algorithm :(
